eclair-core/src/test/scala/fr/acinq/eclair/channel/states/e/NormalSplicesStateSpec.scala

```diff
@@ -0,0 +1,528 @@
+    awaitCond(bob.stateData.asInstanceOf[DATA_NORMAL].commitments.latest.localFundingStatus.asInstanceOf[DualFundedUnconfirmedFundingTx].sharedTx.isInstanceOf[FullySignedSharedTransaction])
+  }
+
+  test("recv CMD_SPLICE (splice-in)") { f =>
+    import f._
+
+    val initialState = alice.stateData.asInstanceOf[DATA_NORMAL]
+    assert(initialState.commitments.latest.capacity == 1_500_000.sat)
+    assert(initialState.commitments.latest.localCommit.spec.toLocal == 800_000_000.msat)
+    assert(initialState.commitments.latest.remoteCommit.spec.toLocal == 700_000_000.msat)
+
+    initiateSplice(f, spliceIn_opt = Some(SpliceIn(500_000 sat)))
+
+    assert(alice.stateData.asInstanceOf[DATA_NORMAL].commitments.latest.capacity == 2_000_000.sat)
+    assert(alice.stateData.asInstanceOf[DATA_NORMAL].commitments.latest.localCommit.spec.toLocal == 1_300_000_000.msat)
+    assert(alice.stateData.asInstanceOf[DATA_NORMAL].commitments.latest.remoteCommit.spec.toLocal == 700_000_000.msat)
+  }
+
+  test("recv CMD_SPLICE (splice-out)") { f =>
+    import f._
+
+    val initialState = alice.stateData.asInstanceOf[DATA_NORMAL]
+    assert(initialState.commitments.latest.capacity == 1_500_000.sat)
+    assert(initialState.commitments.latest.localCommit.spec.toLocal == 800_000_000.msat)
+    assert(initialState.commitments.latest.remoteCommit.spec.toLocal == 700_000_000.msat)
+
+    initiateSplice(f, spliceOut_opt = Some(SpliceOut(100_000 sat, ByteVector32.Zeroes)))
+
+//    val feerate = TestConstants.Alice.nodeParams.onChainFeeConf.feeEstimator.getFeeratePerKw(TestConstants.Alice.nodeParams.onChainFeeConf.feeTargets.fundingBlockTarget)
+//    val miningFee = Transactions.weight2fee(feerate, 600)
+//    assert(alice.stateData.asInstanceOf[DATA_NORMAL].commitments.latest.capacity == 1_400_000.sat - miningFee)
+//    assert(alice.stateData.asInstanceOf[DATA_NORMAL].commitments.latest.localCommit.spec.toLocal == 700_000_000.msat - miningFee)
+//    assert(alice.stateData.asInstanceOf[DATA_NORMAL].commitments.latest.remoteCommit.spec.toLocal == 700_000_000.msat)
+  }
+
+  test("recv WatchFundingConfirmedTriggered on splice tx", Tag(NoMaxHtlcValueInFlight)) { f =>
+    import f._
+
+    val sender = TestProbe()
+    // command for a large payment (larger than local balance pre-slice)
+    val cmd = CMD_ADD_HTLC(sender.ref, 1_000_000_000 msat, randomBytes32(), CltvExpiryDelta(144).toCltvExpiry(currentBlockHeight), TestConstants.emptyOnionPacket, None, localOrigin(sender.ref))
+    // first attempt at payment fails (not enough balance)
+    alice ! cmd
+    sender.expectMsgType[RES_ADD_FAILED[_]]
+    alice2bob.expectNoMessage()
+
+    initiateSplice(f, spliceIn_opt = Some(SpliceIn(500_000 sat)))
+    alice2blockchain.expectMsgType[WatchFundingConfirmed]
+    alice2blockchain.expectNoMessage()
```

Comment:
Caught a bug! We cannot prune _active_ commitments... Fixed in 460f74649.

Reply:
Rolled back the "fix" in 817b5efb1, it broke something else. But the bug is still there, which is why that test isn't passing.

Reply:
> We must not prune active commitments

I think we do: I think we **must** prune active commitments (because once a commitment with a higher `fundingIndex` confirms, nothing bad can happen on the previous ones), but once my other comments around force-closing have been addressed. I believe the handlers for `WatchFundingSpentTriggered` should do the following:

- first check whether the spending transaction is one of our funding transactions by looking into `all` commitments: if that's the case, we can simply ignore that watch
- then try to identify which funding tx has been spent by looking at the `txIn` of the spending transaction and comparing it to our `all` commitments
  - if we cannot find a match, this is a transaction spending a commitment we already pruned: we can simply ignore it, this is because watches are unordered but the spending transaction has to be a pruned funding transaction
  - if we find a match, then we check which type of commitment is spending the funding tx (local, remote, next-remote, revoked)

Reply:
> once a commitment with a higher fundingIndex confirms, nothing bad can happen on the previous ones

I agree that nothing bad can happen in terms of spending the corresponding funding, but that's not all there is to it: if our counterparty has not yet seen the confirmation, then we will have a different set of active commitments and a force close will happen.

Reply:
That's a very good point, you're right! I need to think about this more, this is definitely worth a detailed comment in the code as this is subtle.

Reply:
We will end up with a call to `.pruneCommitments()` in both `updateLocalFundingStatus` _and_ `updateRemoteFundingStatus()`, which, when you think about it, makes sense.

![image](https://user-images.githubusercontent.com/2028222/221586929-aba78ad4-fa78-4e9e-ab05-dc35eb19f245.png)

Reply:
Okay, this is turning a bit rabbit-holey. I think this dual-funding rbf test is wrong:
https://github.com/ACINQ/eclair/blob/817b5efb15701f18c21fef1f0ba16392c629dddc/eclair-core/src/test/scala/fr/acinq/eclair/channel/states/c/WaitForDualFundingConfirmedStateSpec.scala#L303-L304

Because it assumes we have already pruned rbd commitments, which is wrong because in that test we haven't yet received the `channel_ready` from the peer. To be clear, for the initial funding transaction, it works because we are not _using_ the commitments anyway. But I think for consistency we should treat initial and splice funding txs the same. WDYT?

Reply:
I think I have a something that works in cc08b0e28. Small change, but significant impact...

Reply:
I think there is a small issue in `pruneCommitments`: even though we should only prune `inactive` commitments, it should look for the latest confirmed commitment in `all`, not just in `inactive`, because if we have an `active` commitment that is confirmed, we want to automatically prune all `inactive` commitments, whereas the current code wouldn't take it into account.

In `deactivateCommitments`, I'm really not convinced about the change to `<= lastLocked.fundingTxIndex`. Deactivating RBF attempts is fundamentally different from deactivating commitments with a strictly smaller index, where progress is linear. I think we should keep using `< lastLocked.fundingTxIndex` (and I don't think you needed to make that change to get anything to work since we don't yet do RBF on 0-conf txs?).

Reply:
> To be clear, I think this is only an optimization, right?

Since we don't replay `watch-confirmed`, if we don't do this there are times where we will never prune until another splice happens (which could be weeks later). So while this is indeed an optimization, I think it's potentially an important one?

For example if we have `tx0 -> tx1 -> tx2` (where the number indicates the funding index) all unconfirmed, where `tx2` is active and the other two are inactive. They all confirm in the same block, and we receive the watch event for `tx2` first: with the existing code, that will be a no-op. Then we receive the event for `tx1` and prune `tx0`. But `tx1` stays in our `inactive` commitment and won't be cleaned up (because now that `tx2` is confirmed, we don't set a `watch-confirmed` on it anymore).

The only time when it will eventually be cleaned up is when we add a `tx3` AND a `tx4` and the `tx3` confirms while inactive. I think this shows a (small) issue with the model. And conceptually, we really should look at the latest confirmed in both active and inactive, I don't see a reason to restrict it to `inactive` here?

> Do you think it's correct but unnecessary, or plain wrong?

I think it's wrong: I think we cannot deactivate commitments with the same `fundingIndex`, because they could end up being the one that confirms (even though everyone is acting honestly) and if that happens, the state will be completely messed up...

Reply:
You are right, reproduced and fixed the issue in d9ff8ac50.